File: web/src/lib/device/fx.ts
import { z } from "zod";
import type { Currency, Fx } from "../types";
import { addDays, currencies, today } from "../domain";
import type { Engine } from "./engine";
export const cutoff = () => addDays(today(), -1);
export interface RateSnapshot {
  date: string;
  source: string;
  rates: Record<Currency, number>;
}
export function validateRates(
  date: string,
  rates: Record<string, number>,
  source: string,
  target = cutoff(),
): RateSnapshot {
  if (
    !z.iso.date().safeParse(date).success ||
    date > target ||
    Object.keys(rates).length !== 6 ||
    !currencies.every((c) => Number.isFinite(rates[c]) && rates[c] > 0) ||
    rates.USD !== 1
  )
    throw new Error("匯率來源未提供完整的 D-1 快照。");
  return { date, rates: rates as Record<Currency, number>, source };
}
export async function fetchRates(): Promise<RateSnapshot> {
  const target = cutoff();
  try {
    const response = await fetch(
      `https://api.frankfurter.dev/v2/rates?${new URLSearchParams({ base: "USD", quotes: currencies.filter((c) => c !== "USD").join(","), date: target })}`,
      { signal: AbortSignal.timeout(8000) },
    );
    if (!response.ok) throw new Error();
    const records = z
      .array(
        z.object({
          date: z.iso.date(),
          base: z.literal("USD"),
          quote: z.string(),
          rate: z.number().positive(),
        }),
      )
      .length(5)
      .parse(await response.json());
    if (new Set(records.map((r) => r.date)).size !== 1) throw new Error();
    return validateRates(
      records[0].date,
      { USD: 1, ...Object.fromEntries(records.map((r) => [r.quote, r.rate])) },
      "frankfurter.dev/v2",
      target,
    );
  } catch {
    const response = await fetch("https://open.er-api.com/v6/latest/USD", {
      signal: AbortSignal.timeout(8000),
    });
    if (!response.ok)
      throw new Error("目前無法更新匯率，保留最後一次完整快照。");
    const body = z
      .object({
        result: z.literal("success"),
        time_last_update_unix: z.number(),
        rates: z.record(z.string(), z.number()),
      })
      .parse(await response.json());
    return validateRates(
      new Date(body.time_last_update_unix * 1000).toISOString().slice(0, 10),
      Object.fromEntries(currencies.map((c) => [c, body.rates[c]])),
      "open.er-api.com",
      target,
    );
  }
}
export async function fxInfo(e: Engine): Promise<Fx> {
  const [{ date }] = await e.rows<{ date: string | null }>(
    `select max(rate_date) date from (select rate_date from dim_fx_rate where base_ccy='USD' and rate_date<=? and source<>'seed' and quote_ccy in ('USD','EUR','GBP','JPY','TWD','KRW') and rate>0 group by rate_date having count(distinct quote_ccy)=6)`,
    [cutoff()],
  );
  const rows = date
    ? await e.rows<{ quote_ccy: Currency; rate: number; source: string }>(
        "select quote_ccy,rate,source from dim_fx_rate where base_ccy='USD' and rate_date=?",
        [date],
      )
    : [];
  return {
    date,
    source: rows[0]?.source ?? null,
    target_date: cutoff(),
    rates: Object.fromEntries(rows.map((r) => [r.quote_ccy, r.rate])),
  };
}
export async function rate(e: Engine, from: Currency, to: Currency) {
  if (from === to) return { factor: 1, date: null };
  const fx = await fxInfo(e),
    a = fx.rates[from],
    b = fx.rates[to];
  if (!fx.date || !a || !b)
    throw new Error("還沒有可用匯率。請先更新匯率，或以本位幣記帳。");
  return { factor: b / a, date: fx.date };
}
export async function storeRates(e: Engine, snapshot: RateSnapshot) {
  for (const [ccy, value] of Object.entries(snapshot.rates))
    await e.exec(
      "insert or ignore into dim_fx_rate(rate_date,base_ccy,quote_ccy,rate,source) values (?,'USD',?,?,?)",
      [snapshot.date, ccy, value, snapshot.source],
    );
}

File: core/fx.py
"""Immutable, complete FX snapshots with truthful provider dates."""
import datetime as dt
from dataclasses import dataclass
from decimal import Decimal, InvalidOperation, ROUND_HALF_UP
from zoneinfo import ZoneInfo
import requests

BASE = "USD"
HOME_CURRENCIES = ["TWD", "USD", "EUR", "GBP"]
SPEND_CURRENCIES = ["JPY", "TWD", "USD", "EUR", "KRW", "GBP"]
SYMBOL = {"USD": "$", "EUR": "€", "GBP": "£", "TWD": "NT$", "JPY": "¥", "KRW": "₩"}
CURRENCIES = set(HOME_CURRENCIES + SPEND_CURRENCIES)

class RateUnavailable(ValueError):
    pass

def today():
    return dt.datetime.now(ZoneInfo("Asia/Taipei")).date()

def d_minus_1():
    return today() - dt.timedelta(days=1)

def decimal(value):
    try:
        result = Decimal(str(value))
    except (InvalidOperation, ValueError, TypeError):
        raise ValueError("Enter a valid amount.") from None
    if not result.is_finite():
        raise ValueError("Enter a finite amount.")
    return result

@dataclass(frozen=True)
class Snapshot:
    date: dt.date
    rates: dict
    source: str

@dataclass(frozen=True)
class RateStatus:
    date: dt.date | None
    source: str | None
    error: str | None = None

def validate_snapshot(date, rates, source, target):
    if date > target or set(rates) != CURRENCIES:
        raise RateUnavailable("Provider did not return a complete snapshot at or before D-1.")
    parsed = {q: decimal(v) for q, v in rates.items()}
    if any(v <= 0 for v in parsed.values()) or parsed[BASE] != 1:
        raise RateUnavailable("Provider returned invalid rates.")
    return Snapshot(date, parsed, source)

def fetch_rates(target=None):
    target = target or d_minus_1()
    errors = []
    try:
        response = requests.get("https://api.frankfurter.dev/v2/rates", params={
            "base": BASE, "quotes": ",".join(sorted(CURRENCIES - {BASE})), "date": target.isoformat()
        }, timeout=8)
        response.raise_for_status()
        rows = response.json()
        dates = {dt.date.fromisoformat(r["date"]) for r in rows}
        if len(dates) != 1 or any(r["base"] != BASE for r in rows):
            raise RateUnavailable("Rates have inconsistent dates or bases.")
        rates = {r["quote"]: r["rate"] for r in rows}
        rates[BASE] = 1
        return validate_snapshot(dates.pop(), rates, "frankfurter.dev/v2", target)
    except (requests.RequestException, ValueError, KeyError, TypeError) as exc:
        errors.append(type(exc).__name__)
    try:
        response = requests.get(f"https://open.er-api.com/v6/latest/{BASE}", timeout=8)
        response.raise_for_status()
        data = response.json()
        if data.get("result") != "success":
            raise RateUnavailable("Provider reported failure.")
        date = dt.datetime.fromtimestamp(data["time_last_update_unix"], dt.timezone.utc).date()
        return validate_snapshot(date, {q: data["rates"][q] for q in CURRENCIES}, "open.er-api.com", target)
    except (requests.RequestException, ValueError, KeyError, TypeError, OverflowError) as exc:
        errors.append(type(exc).__name__)
    raise RateUnavailable("FX sync unavailable; keeping the last complete snapshot. " + "/".join(errors))

def latest_rate_date(con, on=None):
    row = con.execute("""select max(rate_date) from (
        select rate_date from dim_fx_rate where base_ccy=? and rate_date<=?
        and source<>'seed' and quote_ccy in ('USD','EUR','GBP','JPY','TWD','KRW')
        and rate>0 group by rate_date having count(distinct quote_ccy)=6)""",
        [BASE, on or d_minus_1()]).fetchone()
    return row[0]

def rate_status(con, error=None):
    date = latest_rate_date(con)
    source = con.execute("select min(source) from dim_fx_rate where rate_date=?", [date]).fetchone()[0] if date else None
    return RateStatus(date, source, error)

def ensure_rates(con, force=False):
    from core import db
    with db.LOCK:
        have = latest_rate_date(con)
    if not force and have == d_minus_1():
        with db.LOCK:
            return rate_status(con)
    try:
        snapshot = fetch_rates()
        with db.transaction(con):
            con.executemany("""insert or ignore into dim_fx_rate
                (rate_date,base_ccy,quote_ccy,rate,source) values (?,?,?,?,?)""",
                [(snapshot.date, BASE, q, float(v), snapshot.source) for q, v in snapshot.rates.items()])
        with db.LOCK:
            return rate_status(con)
    except RateUnavailable as exc:
        with db.LOCK:
            return rate_status(con, str(exc))

def factor(con, from_ccy, to_ccy, on=None):
    if from_ccy == to_ccy:
        return Decimal(1), None
    date = latest_rate_date(con, on)
    if date is None:
        raise RateUnavailable("No complete FX snapshot available. Refresh rates before saving foreign-currency expenses.")
    rates = dict(con.execute("select quote_ccy,rate from dim_fx_rate where base_ccy=? and rate_date=?", [BASE, date]).fetchall())
    if from_ccy not in rates or to_ccy not in rates:
        raise RateUnavailable("The selected currency is unavailable in this snapshot.")
    return decimal(rates[to_ccy]) / decimal(rates[from_ccy]), date

def convert(con, amount, from_ccy, to_ccy, on=None):
    return decimal(amount) * factor(con, from_ccy, to_ccy, on)[0]

def fmt(amount, ccy):
    try:
        value = decimal(amount)
    except ValueError:
        return "—"
    places = 0 if ccy in ("JPY", "KRW") else 2
    rounded = value.quantize(Decimal(1).scaleb(-places), rounding=ROUND_HALF_UP)
    return f"{SYMBOL.get(ccy, ccy + ' ')}{rounded:,.{places}f}"
